pkg/vsphere/vsan/synced_cache.go

```diff
@@ -0,0 +1,439 @@
+	}
+}
+
+func (c *syncedDomCache) closeChannels() {
+	defer trace.End(trace.Begin(""))
+	close(c.refresh)
+	close(c.stop)
+	close(c.doneOnce)
+	close(c.waitCurrentRefresh)
+	c.m.Lock()
+	log.Debugf("Finish all waitgroups")
+	for i := range c.refreshings {
+		if len(c.refreshings[i]) == 0 {
+			continue
+		}
+		c.doneWaitGroups(c.refreshings[i])
+	}
+}
+
+func (c *syncedDomCache) doneWaitGroups(wgs []*sync.WaitGroup) {
+	defer trace.End(trace.Begin(fmt.Sprintf("%#v waitgroups is done", wgs)))
+	for i := range wgs {
+		wgs[i].Done()
+	}
+}
+
+func (c *syncedDomCache) datastoreDom(ds string) DSDomCache {
+	c.m.RLock()
+	defer c.m.RUnlock()
+	return c.dsMap[ds]
+}
+
+// AddDomCache create dom cache for Datastore
+func (c *syncedDomCache) AddDomCache(ctx context.Context, ds *object.Datastore) error {
+	if ds == nil {
+		defer trace.End(trace.Begin(fmt.Sprintf("datastore is empty")))
+		return nil
+	}
+	defer trace.End(trace.Begin(ds.Reference().String()))
+
+	ref := ds.Reference().String()
+	if dsc := c.datastoreDom(ref); dsc != nil {
+		log.Debugf("Dom cache for datastore %s already exists", ds.InventoryPath)
+		return nil
+	}
+
+	if dsType, _ := ds.Type(ctx); dsType != types.HostFileSystemVolumeFileSystemTypeVsan {
+		log.Debugf("datastore %s is not vsan, no need to build vsan dom cache", ds.InventoryPath)
+		return nil
+	}
+
+	hosts, err := ds.AttachedHosts(ctx)
+	if err != nil {
+		err = errors.Errorf("failed to get attached hosts for datastore %s: %s", ds.InventoryPath, err)
+		log.Error(err)
+		return err
+	}
+	if len(hosts) == 0 {
+		err = errors.Errorf("no host attached to datastore %s", ds.InventoryPath)
+		log.Error(err)
+		return err
+	}
+
+	// vsan object can be query and removed from any host whatever the owner is, so here just get the first host
+	hc := hosts[0].ConfigManager()
+	var h mo.HostSystem
+
+	err = hc.Properties(ctx, hc.Reference(), []string{"configManager.vsanInternalSystem"}, &h)
+	if err != nil {
+		err = errors.Errorf("failed to get vsanInternalSystem for host %s: %s", hosts[0].Reference(), err)
+		log.Error(err)
+		return err
+	}
+
+	if h.ConfigManager.VsanInternalSystem == nil {
+		err = errors.Errorf("vsanInternalSystem of host %s is empty", hosts[0].Reference())
+		log.Error(err)
+		return err
+	}
+
+	hvis := NewHostVsanInternalSystem(ds.Client(), *h.ConfigManager.VsanInternalSystem)
+	dsc := &vsanDSDomCache{
+		ds:    ds,
+		hvis:  hvis,
+		uuids: make(map[string]string),
+		paths: make(map[string]string),
+		m:     &sync.Mutex{},
+	}
+	c.m.Lock()
+	defer c.m.Unlock()
+	c.dsMap[ds.Reference().String()] = dsc
+	log.Debugf("Dom cache for datastore %s is created, start to refresh", ds.InventoryPath)
+	c.Refresh(ds.Reference().String(), nil)
+	return nil
+}
+
+func (c *syncedDomCache) getNonRefreshingDS(requests []*dsWaitGroup) []*dsWaitGroup {
+	defer trace.End(trace.Begin(fmt.Sprintf("%#v", requests)))
+
+	c.m.Lock()
+	defer c.m.Unlock()
+	var ret []*dsWaitGroup
+	for i := range requests {
```

Comment:
request is an array, so a range can provide you a value that you would not access by index.